File: ProjectEuler/201-250/S205.cs
using System.Collections.Generic;

namespace ProjectEuler
{
    public class S205 : ISolution
    {
        private static readonly List<string> PeterResults = [];
        private static readonly List<string> ColinResults = [];

        public string GetAnswer()
        {
            int[] Peter = new int[37];
            int[] Colin = new int[37];

            GenPeterResults("", 0);
            PeterResults.ForEach(s => Peter[Utils.DigitSum(int.Parse(s))]++);

            GenColinResults("", 0);
            ColinResults.ForEach(s => Colin[Utils.DigitSum(int.Parse(s))]++);

            long numerator = 0;
            for (int p = 1; p < Peter.Length; p++)
            {
                for (int c = 1; c < Colin.Length; c++)
                {
                    if (p > c)
                    {
                        numerator += Peter[p] * Colin[c];
                    }
                }
            }

            return ((double)numerator / 262144 / 46656).ToString("F7");
        }

        private static void GenPeterResults(string num, int nth)
        {
            for (int i = 1; i <= 4; i++)
            {
                if (nth == 8)
                {
                    PeterResults.Add(num + i);
                }
                else
                {
                    GenPeterResults(num + i, nth + 1);
                }
            }
        }

        private static void GenColinResults(string num, int nth)
        {
            for (int i = 1; i <= 6; i++)
            {
                if (nth == 5)
                {
                    ColinResults.Add(num + i);
                }
                else
                {
                    GenColinResults(num + i, nth + 1);
                }
            }
        }
    }
}

File: src/201-250/S205.cc
#include "S205.h"
#include "../Utils.h"
#include <iomanip>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

namespace ProjectEuler
{
std::string S205::GetAnswer()
{
	PeterResults.clear();
	ColinResults.clear();

	std::vector<int> Peter(37, 0); // Max sum 4*9 = 36.
	std::vector<int> Colin(37, 0); // Max sum 6*6 = 36.

	GenPeterResults("", 0);
	for (const auto &s : PeterResults)
	{
		long long sum = 0;
		for (char c : s)
		{
			sum += (c - '0');
		}
		if (sum < Peter.size())
		{
			Peter[sum]++;
		}
	}

	GenColinResults("", 0);
	for (const auto &s : ColinResults)
	{
		long long sum = 0;
		for (char c : s)
		{
			sum += (c - '0');
		}
		if (sum < Colin.size())
		{
			Colin[sum]++;
		}
	}

	long long numerator = 0;
	for (size_t p = 1; p < Peter.size(); p++)
	{
		for (size_t c = 1; c < Colin.size(); c++)
		{
			if (p > c)
			{
				numerator += (long long)Peter[p] * Colin[c];
			}
		}
	}

	double result = (double)numerator / 262144.0 / 46656.0;

	std::stringstream ss;
	ss << std::fixed << std::setprecision(7) << result;
	return ss.str();
}

void S205::GenPeterResults(std::string num, int nth)
{
	if (nth == 9)
	{
		PeterResults.push_back(num);
		return;
	}

	for (int i = 1; i <= 4; i++)
	{
		GenPeterResults(num + std::to_string(i), nth + 1);
	}
}

void S205::GenColinResults(std::string num, int nth)
{
	if (nth == 6)
	{
		ColinResults.push_back(num);
		return;
	}

	for (int i = 1; i <= 6; i++)
	{
		GenColinResults(num + std::to_string(i), nth + 1);
	}
}

} // namespace ProjectEuler
